dolphin/db/sqlalchemy/api.py

```diff
@@ -136,23 +150,37 @@ def storage_get(context, storage_id):
 
 def storage_get_all(context, marker=None, limit=None, sort_keys=None,
                     sort_dirs=None, filters=None, offset=None):
-    """Retrieves all storage devices."""
-
     this_session = get_session()
     this_session.begin()
-    # TODO: need to handle all input parameters
-    all_storages = this_session.query(Storage).all()
-    return all_storages
-
-
-def volume_create(context, values):
-    """Create a volume from the values dictionary."""
-    return NotImplemented
-
+    if not sort_keys:
+        sort_keys= ['created_at']
+    if not sort_dirs:
+        sort_dirs = ['desc']
+    this_session = get_session()
+    this_session.begin()
+    query = this_session.query(models.Storage)
+
+    if filters:
+
+        for attr, value in filters.items():
+            query = query.filter(getattr(models.Storage, attr).like("%%%s%%" % value))
+    try:
+        for (sort_key,sort_dir) in zip(sort_keys,sort_dirs ):
+            query = apply_sorting(models.Storage, query, sort_key, sort_dir)
+    except AttributeError:
+        msg = "Wrong sorting keys provided - '%s'." % sort_keys
+        raise exception.InvalidInput(reason=msg)
+
+    if limit:
+        query = query.limit(limit)
+
+    # Returns list of storages  that satisfy filters.
+    storages_all = query.all()
+    if storages_all:
+        return storages_all
+    else:
+        raise exception.NotFound()
```

Comment:
For xxx_get_all APIs, if nothing found, that's a normal scenario, so no need to raise exception here. Only xxx_get APIs need to raise NotFound exception while specific object not found.

Reply:
Done